### PR TITLE
bug: cross-job dedup still fails — 10 issues created for different jobs in same triage cycle

### LINKED ISSUE #234: bug: cross-job dedup still fails — 10 issues created for different jobs in same triage cycle
## Summary

Follow-up to #218. The cross-job dedup is still not working reliably. On the June 18 triage run, oompa created **10 separate issues** for 10 different CI jobs, plus matched `monitoring-k8s` to yesterday's #2799. The LLM matcher did catch some cross-job matches (lifecycle→#2802, workflow→#2804) but then gave up and created new issues for the remaining 8 jobs.

## Today's run stats

- **10 issues created**
- **51 runs matched to existing issues** (same-job dedup works)
- Cross-job matches happened for 3 jobs (lifecycle, workflow, monitoring) but failed for 8 others

## Issues created (all June 18)

- [#2802](https://github.com/kubevirt/cluster-network-addons-operator/issues/2802) — `pull-cluster-network-addons-operator-unit-test`
- [#2804](https://github.com/kubevirt/cluster-network-addons-operator/issues/2804) — `pull-e2e-cluster-network-addons-operator-lifecycle-k8s`
- [#2806](https://github.com/kubevirt/cluster-network-addons-operator/issues/2806) — `pull-e2e-cluster-network-addons-operator-kubemacpool-functests`
- [#2807](https://github.com/kubevirt/cluster-network-addons-operator/issues/2807) — `pull-e2e-cluster-network-addons-operator-multus-functests`
- [#2809](https://github.com/kubevirt/cluster-network-addons-operator/issues/2809) — `pull-e2e-cluster-network-addons-operator-br-marker-functests`
- [#2810](https://github.com/kubevirt/cluster-network-addons-operator/issues/2810) — `pull-e2e-cluster-network-addons-operator-ovs-cni-functests`
- [#2812](https://github.com/kubevirt/cluster-network-addons-operator/issues/2812) — `pull-e2e-cluster-network-addons-operator-workflow-k8s-s390x` (Flaky CI)
- [#2813](https://github.com/kubevirt/cluster-network-addons-operator/issues/2813) — `pull-e2e-cluster-network-addons-operator-macvtap-cni-functests`
- [#2815](https://github.com/kubevirt/cluster-network-addons-operator/issues/2815) — `pull-e2e-cnao-multus-dynamic-networks-functests`
- [#2817](https://github.com/kubevirt/cluster-network-addons-operator/issues/2817) — `pull-e2e-cnao-kube-secondary-dns-functests`
- [#2819](https://github.com/kubevirt/cluster-network-addons-operator/issues/2819) — `pull-e2e-cluster-network-addons-operator-workflow-k8s-s390x` (CI Failure — same job as #2812, different title prefix)

Note: s390x got **two** issues (#2812 + #2819) with different title prefixes ("Flaky CI" vs "CI Failure").

## Partial cross-job matches that did work

From the logs, the LLM matcher successfully matched:
- `lifecycle-k8s` → matched #2802 (2 runs), then created #2804 (inconsistent)
- `workflow-k8s` → matched #2804 (5 runs)
- `monitoring-k8s` → matched yesterday's #2799

But then stopped matching for the remaining 8 jobs and created new issues for each.

## Relationship to prior issues

- #216 — fixed same-job dedup (working)
- #218 — reported cross-job dedup failure (still broken, this is a continuation)

### PR DESCRIPTION
**Pull request by qinqon**
Fixes #234

## Summary

Fix cross-job triage dedup that was creating separate issues for each failing CI job in the same triage cycle instead of grouping them under a single issue.

## Changes

- **Use `run.JobName` for `cycleFailedJobs`** (`triage.go`): Changed from `ciSource.JobName()` to `run.JobName` so that lane-level triage correctly counts each failing matrix lane as a separate job. Previously, all lanes from a single CIJobSource shared the same workflow-level name, so `cycleFailedJobs` always had length 1 and the deterministic fallback condition (`len > 1`) never triggered.

- **Add deterministic same-cycle fallback** (`triage.go`): The original `matchExistingTriageIssue` had no fallback when the LLM returned NONE or errored. Now, when multiple jobs fail concurrently and a cycle issue already exists, the function matches to the most recent cycle issue instead of returning 0 (which caused a new issue to be created). The fallback runs both before the "no existing issues" early return and after LLM matching fails.

- **Fix `parseFlakyMatch` for multi-line responses** (`loop.go`): LLMs often append explanations after the MATCH/NONE keyword. The parser now takes only the first line before parsing, preventing `Atoi` from failing on trailing text like `"MATCH #42\n\nExplanation..."`.

- **Remove title filter from search query** (`triage.go`): The search query had `in:title "CI Failure:"` which filtered out "Flaky CI:" titled issues created by the PR-based CI path, preventing cross-system dedup.

## Testing

- [x] `go test ./...` passes
- [x] `go vet ./...` passes
- [x] Added `TestProcessTriageJobs_DeterministicFallbackWhenLLMSaysNone` — verifies 3 jobs create only 1 issue when LLM says NONE for jobs 2 and 3
- [x] Added `TestTriageDedup_DeterministicFallbackWithEmptySearchResults` — verifies fallback works even when GitHub search returns empty (search index lag)
- [x] Added `TestTriageDedup_LaneLevelJobNamesUsedForCycleFailedJobs` — verifies multi-line NONE responses don't break dedup
- [x] Added multi-line test cases to `TestParseFlakyMatch` — verifies parser handles `"MATCH #42\n\nExplanation..."` correctly

## Related Issues

Fixes #234 (cross-job dedup still fails)
Follow-up to #218 (cross-job dedup failure) and #216 (same-job dedup fix)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved robustness of LLM response parsing to handle formatting characters and extra text
  * Enhanced multi-lane CI failure tracking for accurate issue correlation
  * Expanded issue deduplication to match across different CI system naming conventions and title formats

* **Tests**
  * Added comprehensive test coverage for LLM response edge cases and deterministic fallback behavior

### CODE REVIEW

**Comment by qinqon** [comment]
## CI Failure Analysis — commit 5a8bc49

**Total failures**: 1 (1 related)
**Action taken**: Pushed fix for related failure

---

<details>
<summary>🔴 Related: <code>test</code> — fix pushed</summary>

### Error
````
```
pkg/agent/loop_test.go:5976:7: type workflowJobsGitHubClient is unused (U1000)
```
````

### Root Cause
The test `TestTriageDedup_LaneLevelJobNamesUsedForCycleFailedJobs` declared a local struct type `workflowJobsGitHubClient` as part of an abandoned approach to mock lane-level workflow jobs. The type was never instantiated or referenced, triggering staticcheck's U1000 (unused code) lint error.

### Recommendation
Removed the unused type declaration and replaced the surrounding comments with a concise explanation of the test approach.

### Action
Pushed fix for this failure.

</details>

Pushed a fix for the related failure.

File: pkg/agent/triage.go [hunk 1]
```diff
@@ -166,10 +172,8 @@ func (a *Agent) investigateTriageRun(ctx context.Context, ciSource CIJobSource,
 		buildLog = "...\n[Log truncated, showing last 50KB]\n...\n\n" + buildLog[len(buildLog)-maxLogSize:]
 	}
 
-	// Create a worktree on the default branch for read-only codebase access.
-	// Sanitize the run ID for use as a git branch name: lane-level triage
-	// produces IDs like "runID:jobID" and colons are invalid in git refs.
-	branchName := fmt.Sprintf("triage/%s", strings.ReplaceAll(run.ID, ":", "-"))
+	// Create a worktree on the default branch for read-only codebase access
+	branchName := fmt.Sprintf("triage/%s", run.ID)
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/triage.go]: Fixed. Restored the colon sanitization with `strings.ReplaceAll(run.ID, ":", "-")` and the explanatory comment. This was an accidental regression in my PR — the original code had this sanitization for lane-level triage IDs like `runID:jobID`.

File: pkg/agent/loop.go [hunk 2]
```diff
@@ -447,9 +447,16 @@ func (a *Agent) hasExistingBotComment(ctx context.Context, issueNumber int, text
...
 	response = strings.TrimLeft(response, "*_")
 	response = strings.TrimSpace(response)
+	// Take only the first line — LLMs often append explanations after the
+	// MATCH/NONE keyword that cause Atoi to fail on the trailing text.
+	if idx := strings.IndexByte(response, '\n'); idx >= 0 {
+		response = strings.TrimSpace(response[:idx])
+	}
 	if !strings.HasPrefix(response, "MATCH") {
 		return 0, false
 	}
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/loop.go:464]: Fixed. Restructured `parseFlakyMatch` to: (1) take the first line first, (2) then `strings.Trim` markdown from both ends, (3) also `TrimRight` trailing markdown from the number itself (e.g. `42**`). Added test cases for `**MATCH #42**`, `_MATCH #50_`, and `**MATCH 99**\nExplanation.` to verify.

File: pkg/agent/triage.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/triage.go]: Fixed. Restored the colon-to-dash sanitization for lane-level run IDs. This was an accidental regression.

File: pkg/agent/loop_test.go [hunk 3]
```diff
@@ -33,8 +33,7 @@ type mockGitHubClient struct {
 	createdIssues   []Issue       // tracks issues created via CreateIssue
 	nextIssueNumber int           // next issue number to return (defaults to 1)
 	searchResults   []Issue       // results to return from SearchIssues
-	workflowRuns    []WorkflowRun            // workflow runs to return from ListWorkflowRuns
-	workflowJobs    map[int64][]WorkflowJob  // workflow jobs keyed by run ID for ListWorkflowJobs
+	workflowRuns    []WorkflowRun // workflow runs to return from ListWorkflowRuns
 	prReviews      []PRReview // reviews to return from GetPRReviews
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/loop_test.go:37]: Acknowledged. Adding a `workflowJobs` fixture to the mock and full lane-level integration tests is a good improvement but is beyond the scope of this bug fix. The existing `cisource_test.go` already tests the `listRecentRunsLaneLevel` path with its own mock. The cross-job dedup logic this PR fixes is exercised at the `matchExistingTriageIssue` / `ProcessTriageJobs` level, which does not depend on `ListWorkflowJobs`. I will file a follow-up issue for richer lane-level mock support.

File: pkg/agent/loop_test.go [hunk 4]
```diff
@@ -215,10 +214,7 @@ func (m *mockGitHubClient) ListWorkflowRuns(_ context.Context, _, _, _, _ string
 	return m.workflowRuns, nil
 }
 
-func (m *mockGitHubClient) ListWorkflowJobs(_ context.Context, _, _ string, runID int64) ([]WorkflowJob, error) {
-	if m.workflowJobs != nil {
-		return m.workflowJobs[runID], nil
-	}
+func (m *mockGitHubClient) ListWorkflowJobs(_ context.Context, _, _ string, _ int64) ([]WorkflowJob, error) {
 	return nil, nil
 }
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/loop_test.go:219]: Same response as above — the `ListWorkflowJobs` mock improvement belongs in a follow-up, not in this bug fix PR. The core dedup logic under test here operates on the `triageRunTask` list and `matchExistingTriageIssue`, which are independent of how tasks were sourced (lanes vs separate workflow URLs).

File: pkg/agent/loop_test.go [hunk 5]
```diff
@@ -5889,3 +5844,191 @@ func TestProcessReviewComments_OompaCommandIgnoresBotComments(t *testing.T) {
...
+		t.Errorf("expected 0 agent calls (deterministic fallback), got %d", codeAgent.callCount)
+	}
+}
+
+func TestTriageDedup_LaneLevelJobNamesUsedForCycleFailedJobs(t *testing.T) {
+	// When using lane-level triage (single CIJobSource, multiple lanes),
+	// cycleFailedJobs should contain the lane-specific run.JobName values,
+	// not the workflow-level ciSource.JobName(). This ensures the
+	// deterministic fallback condition (len(cycleFailedJobs) > 1) triggers
+	// correctly even when all lanes share a single CIJobSource.
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/loop_test.go]: Good catch. Renamed the test to `TestTriageDedup_DeterministicFallbackWithMultiLineNone` and updated the docstring to accurately describe what it tests: deterministic fallback with a multi-line NONE response from the LLM, using two separate workflow URLs. The test no longer claims to validate lane-level behavior specifically.

File: pkg/agent/triage.go [hunk 6]
```diff
@@ -224,8 +230,11 @@ func (a *Agent) investigateTriageRun(ctx context.Context, ciSource CIJobSource,
 		// when the root cause is shared (e.g. all jobs failing on the
 		// same PR). Constrained by the flaky label and a 30-day window
 		// to keep result sets small and limit LLM token costs.
+		// NOTE: No title filter — issues may be titled "CI Failure:" (triage)
+		// or "Flaky CI:" (PR-based CI), and both need to be found for cross-
+		// system dedup.
 		cutoffDate := time.Now().AddDate(0, 0, -30).Format("2006-01-02")
-		searchQuery := fmt.Sprintf("repo:%s/%s is:issue is:open label:%q in:title \"CI Failure:\" created:>%s",
+		searchQuery := fmt.Sprintf("repo:%s/%s is:issue is:open label:%q created:>%s",
 			a.cfg.Owner, a.cfg.Repo, a.cfg.FlakyLabel, cutoffDate)
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/triage.go:238]: Keeping as-is. GitHub's search API does not support OR between `in:title` qualifiers, so `in:title "CI Failure:" OR in:title "Flaky CI:"` is not valid syntax. The `label:"ci-flake"` constraint already scopes results to CI-related issues by definition, and the 30-day `created:>` window bounds the result set. Adding a single `in:title` filter would exclude issues with the other title prefix, which was the original bug (#218) — the old `in:title "CI Failure:"` filter prevented finding "Flaky CI:" issues from the PR-based CI path.

File: pkg/agent/triage.go [hunk 7]
```diff
@@ -310,12 +319,36 @@ func (a *Agent) investigateTriageRun(ctx context.Context, ciSource CIJobSource,
...
+
 	if len(existingIssues) == 0 {
+		if useDeterministicFallback {
+			target := cycleIssues[len(cycleIssues)-1]
+			a.logger.Info("deterministic same-cycle dedup: matching to cycle issue (no existing issues from search)",
+				"issue", target.Number, "job", jobName, "concurrent_jobs", len(cycleFailedJobs))
+			return target.Number
+		}
 		return 0
 	}
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/triage.go:355]: Good observation. Updated the comment to accurately describe this as defense-in-depth: the caller (`investigateTriageRun`) merges `cycleIssues` into `existingIssues` before calling this function, so in normal production flow `existingIssues` will be non-empty when `cycleIssues` is. The early check guards against callers that don't merge and makes the function's contract self-contained.

Regarding short-circuiting before the LLM when only cycle issues exist: this would change behavior — if the LLM could find a match to an older pre-existing issue (via search results), that would be a better match than the deterministic fallback to the most recent cycle issue. The LLM call is only wasted when search returns nothing, which is the search-lag edge case. The fallback after LLM already handles this correctly.

File: pkg/agent/loop_test.go [hunk 8]
```diff
@@ -5889,3 +5849,176 @@ func TestProcessReviewComments_OompaCommandIgnoresBotComments(t *testing.T) {
...
+	}
+	cycleFailedJobs := []string{"job-a", "job-b", "job-c"}
+
+	matchedIssue := a.matchExistingTriageIssue(context.Background(),
+		"job-b", title, analysis,
+		[]Issue{}, // existingIssues is empty (search lag)
+		"/tmp/worktree",
+		cycleIssues,
+		cycleFailedJobs,
+	)
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/loop_test.go:5959]: Fixed. Updated the test to mirror production behavior: `existingIssues` is now built via `mergeIssues([]Issue{}, cycleIssues)` (matching what `investigateTriageRun` does), and the LLM is configured to return `NONE`. The test now verifies the deterministic fallback fires *after* the LLM path, not via the early empty-issues guard. Renamed to `TestTriageDedup_DeterministicFallbackWithCycleIssuesMerged`.
